File: app/src/main/java/com/jossy/android/mobilebenchmarkappjava/LocationTestActivity.java
package com.jossy.android.mobilebenchmarkappjava;

import android.Manifest;
import android.content.pm.PackageManager;
import android.location.Location;
import android.os.Bundle;
import android.os.Looper;
import android.util.Log;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.Priority;

public class LocationTestActivity extends AppCompatActivity {
    private static final int PERMISSION_REQUEST_CODE = 123;
    private static final long UPDATE_INTERVAL = 1000L;
    private static final long FASTEST_UPDATE_INTERVAL = 500L;
    private static final int ACCURACY_THRESHOLD = 20; // meters

    private FusedLocationProviderClient fusedLocationClient;
    private LocationCallback locationCallback;
    private TextView locationStatus, accuracyView, timeToFirstFix, coordinates;
    private long startTime;
    private boolean firstFix = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d("LocationTestActivity", "onCreate called");
        setContentView(R.layout.activity_location_test);

        locationStatus = findViewById(R.id.locationStatus);
        accuracyView = findViewById(R.id.accuracyView);
        timeToFirstFix = findViewById(R.id.timeToFirstFix);
        coordinates = findViewById(R.id.coordinates);
        Log.d("LocationTestActivity", "Views initialized");

        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
        setupLocationCallback();

        if (checkPermissions()) {
            startLocationUpdates();
        } else {
            requestPermissions();
        }
    }

    private void setupLocationCallback() {
        locationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(@NonNull LocationResult locationResult) {
                Location location = locationResult.getLastLocation();
                if (location != null) {
                    updateLocationInfo(location);
                }
            }
        };
    }

    private void updateLocationInfo(Location location) {
        float accuracy = location.getAccuracy();
        accuracyView.setText("Accuracy: " + accuracy + " meters");
        coordinates.setText(String.format("Location: %.6f, %.6f", 
                location.getLatitude(), location.getLongitude()));

        if (firstFix) {
            long ttff = System.currentTimeMillis() - startTime;
            timeToFirstFix.setText("Time to first fix: " + ttff + "ms");
            firstFix = false;
            locationStatus.setText("Location acquired!");

            if (accuracy <= ACCURACY_THRESHOLD) {
                stopLocationUpdates();
                locationStatus.setText("Test completed - Good accuracy achieved");
            }
        }
    }

    private boolean checkPermissions() {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) 
                == PackageManager.PERMISSION_GRANTED;
    }

    private void requestPermissions() {
        ActivityCompat.requestPermissions(this,
                new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                PERMISSION_REQUEST_CODE);
    }

    private void startLocationUpdates() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) 
                != PackageManager.PERMISSION_GRANTED) {
            return;
        }

        LocationRequest locationRequest = new LocationRequest.Builder(UPDATE_INTERVAL)
                .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
                .setMinUpdateIntervalMillis(FASTEST_UPDATE_INTERVAL)
                .build();

        startTime = System.currentTimeMillis();
        locationStatus.setText("Requesting location updates...");

        fusedLocationClient.requestLocationUpdates(locationRequest,
                locationCallback, Looper.getMainLooper());
    }

    private void stopLocationUpdates() {
        fusedLocationClient.removeLocationUpdates(locationCallback);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                         @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == PERMISSION_REQUEST_CODE && grantResults.length > 0
                && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            startLocationUpdates();
        } else {
            locationStatus.setText("Location permission denied");
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        stopLocationUpdates();
    }
}

File: app/src/main/java/com/jossy/android/mobilebenchmarkappkotlin/activity/LocationTestActivity.kt
package com.jossy.android.mobilebenchmarkappkotlin.activity

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.google.android.gms.location.*
import com.jossy.android.mobilebenchmarkappkotlin.BenchmarkApplication
import com.jossy.android.mobilebenchmarkappkotlin.R
import com.jossy.android.mobilebenchmarkappkotlin.io.BufferedCsvWriter
import com.jossy.android.mobilebenchmarkappkotlin.model.TestResult
import java.io.File

class LocationTestActivity : AppCompatActivity() {

    private lateinit var statusText: TextView
    private var sampleCount: Int = 10000
    private var outputFile: String? = null
    private var csvWriter: BufferedCsvWriter? = null
    
    private lateinit var fusedLocationClient: FusedLocationProviderClient
    private lateinit var locationCallback: LocationCallback
    private var samplesCollected = 0
    private var startTimeMs = 0L
    private val handler = Handler(Looper.getMainLooper())
    
    private var lastLocationTime = 0L
    private val watchdogRunnable = object : Runnable {
        override fun run() {
            if (System.currentTimeMillis() - lastLocationTime > 15000) { // 15s watchdog
                finishBenchmark(false, "Watchdog timeout - no signal")
            } else {
                handler.postDelayed(this, 1000)
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_location_test)
        
        statusText = findViewById(R.id.locationStatus)

        sampleCount = intent.getIntExtra("sample_count", 10000)
        outputFile = intent.getStringExtra("output_file")
        
        initializeWriter()
        
        fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
        
        startLocationUpdates()
    }

    private fun initializeWriter() {
        outputFile?.let { path ->
            csvWriter = BufferedCsvWriter(File(path), 1000, 64 * 1024)
            csvWriter?.initialize()
        }
    }

    @SuppressLint("MissingPermission")
    private fun startLocationUpdates() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            finishBenchmark(false, "No permission")
            return
        }

        val locationRequest = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 1000)
            .setMinUpdateIntervalMillis(500)
            .build()
            
        startTimeMs = System.currentTimeMillis()
        lastLocationTime = startTimeMs

        locationCallback = object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult) {
                if (samplesCollected >= sampleCount) return
                
                for (location in locationResult.locations) {
                    val now = System.currentTimeMillis()
                    lastLocationTime = now
                    samplesCollected++

                    csvWriter?.write(
                        iteration = samplesCollected,
                        executionTimeMs = 0, // Instant
                        details = "lat=${location.latitude}|lon=${location.longitude}|acc=${location.accuracy}",
                        intervalStartMs = now,
                        intervalDurationMs = 0,
                        cumulativeTimeMs = now - startTimeMs
                    )
                    
                    if (samplesCollected % 10 == 0) {
                        statusText.text = "Samples: $samplesCollected/$sampleCount"
                    }

                    if (samplesCollected >= sampleCount) {
                        finishBenchmark(true, "Completed $samplesCollected samples")
                        return
                    }
                }
            }
        }

        fusedLocationClient.requestLocationUpdates(locationRequest, locationCallback, Looper.getMainLooper())
        handler.postDelayed(watchdogRunnable, 1000)
    }
    
    private fun finishBenchmark(success: Boolean, msg: String) {
        handler.removeCallbacks(watchdogRunnable)
        fusedLocationClient.removeLocationUpdates(locationCallback)
        csvWriter?.close()
        
        val totalTime = System.currentTimeMillis() - startTimeMs
        val result = TestResult(
            "Location Test",
            totalTime,
            msg,
            success
        )
        val intent = Intent()
        intent.putExtra(BenchmarkApplication.RESULT, result)
        setResult(RESULT_OK, intent)
        finish()
    }
}
